File: front/src/app/shared/model/mapper/pessoa-mapper.ts
// pessoa.mapper.ts

import {PessoaDto} from "@/shared/model/dto/pessoa-dto";
import {Pessoa} from "@/shared/model/pessoa";
import {EventoPessoa} from "@/shared/model/evento-pessoa";
import {ProdutoDto} from "@/shared/model/dto/produto-dto";
import {ClienteMapper} from "@/shared/model/mapper/cliente-mapper";
import {StatusEnum} from "@/shared/model/enum/status.enum";

export class PessoaMapper {

    /**
     * Entidade -> DTO (Prepara dados vindo da API para o componente)
     */
    static toDTO(entity: any): PessoaDto {
        if (!entity) return {} as PessoaDto;

        return {
            id: entity.id,
            cliente: ClienteMapper.toDto(entity.cliente),
            nome: entity.nome,
            cpf: entity.cpf,
            telefone: entity.telefone,
            email: entity.email,
            status: StatusEnum.toKey(entity.status),
            endereco: entity.endereco,
            complemento: entity.complemento,
            cidade: entity.cidade,
            estado: entity.estado,
            cep: entity.cep,
            senha: entity.senha,
            version: entity.version
        };
    }

    /**
     * DTO -> Entidade (Prepara dados do formulário para enviar à API)
     */
    static fromDTO(dto: PessoaDto): Pessoa {
        if (!dto) return new Pessoa();

        const pessoa = new Pessoa();

        // Copia todas as propriedades simples que possuem o mesmo nome
        Object.assign(pessoa, dto);

        // Trata a conversão de clienteId (number) para o objeto cliente ({id: number})
        if (dto.cliente) {
            pessoa.cliente = ClienteMapper.fromDto(dto.cliente);
        }
        else {
            pessoa.cliente = undefined;
        }

        return pessoa;
    }

    static toDtoList(entities: any[]): PessoaDto[] {
        if (!entities) return [];
        return entities.map(entity => this.toDTO(entity))
            .filter(dto => !!dto) as PessoaDto[];
    }

    static listFromDto(dtos: PessoaDto[]): Pessoa[] {
        if (!dtos) return [];
        return dtos.map(dto => this.fromDTO(dto)).filter(dto => !!dto) as Pessoa[];
    }
}

File: back/src/main/java/br/eng/eliseu/presente/model/mapper/PessoaMapper.java
package br.eng.eliseu.presente.model.mapper;

import br.eng.eliseu.presente.model.Pessoa;
import br.eng.eliseu.presente.model.dto.PessoaDto;
import org.springframework.stereotype.Component;
import org.springframework.util.ObjectUtils;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Component
public class PessoaMapper {

    /**
     * Entidade -> DTO
     */
    public static PessoaDto toDto(Pessoa pessoa) {
        if (pessoa == null) return null;

        return new PessoaDto(
                pessoa.getId(),
                ClienteMapper.toDto(pessoa.getCliente()),
                pessoa.getNome(),
                pessoa.getCpf(),
                pessoa.getTelefone(),
                pessoa.getEmail(),
                pessoa.getStatus() != null ? pessoa.getStatus() : null,
                pessoa.getEndereco(),
                pessoa.getComplemento(),
                pessoa.getCidade(),
                pessoa.getEstado(),
                pessoa.getCep(),
                pessoa.getSenha(),
                pessoa.getCriadoEm(),
                pessoa.getAlteradoEm(),
                pessoa.getVersion()
        );
    }

    /**
     * DTO -> Entidade
     */
    public static Pessoa fromDto(PessoaDto dto) {
        return fromDto(dto, null);
    }

    public static Pessoa fromDto(PessoaDto dto, Pessoa pessoaExistente) {
        if (dto == null || ObjectUtils.isEmpty(dto)) {
            return null;
        }

        // Criamos uma nova variável final que recebe ou a existente ou uma nova
        final Pessoa pessoaFinal = (pessoaExistente != null) ? pessoaExistente : new Pessoa();

        // Agora a Lambda funciona porque a variável 'pessoa' é final
        Optional.ofNullable(dto.cliente())
                .ifPresent(c -> pessoaFinal.setCliente(ClienteMapper.fromDto(c, pessoaFinal.getCliente())));

        Optional.ofNullable(dto.nome()).ifPresent(pessoaFinal::setNome);
        Optional.ofNullable(dto.cpf()).ifPresent(pessoaFinal::setCpf);
        Optional.ofNullable(dto.telefone()).ifPresent(pessoaFinal::setTelefone);
        Optional.ofNullable(dto.email()).ifPresent(pessoaFinal::setEmail);
        Optional.ofNullable(dto.endereco()).ifPresent(pessoaFinal::setEndereco);
        Optional.ofNullable(dto.complemento()).ifPresent(pessoaFinal::setComplemento);
        Optional.ofNullable(dto.cidade()).ifPresent(pessoaFinal::setCidade);
        Optional.ofNullable(dto.estado()).ifPresent(pessoaFinal::setEstado);
        Optional.ofNullable(dto.cep()).ifPresent(pessoaFinal::setCep);
        Optional.ofNullable(dto.senha()).ifPresent(pessoaFinal::setSenha);
        Optional.ofNullable(dto.status()).ifPresent(pessoaFinal::setStatus);
        Optional.ofNullable(dto.criadoEm()).ifPresent(pessoaFinal::setCriadoEm);
        Optional.ofNullable(dto.alteradoEm()).ifPresent(pessoaFinal::setAlteradoEm);

        return pessoaFinal;
    }

    /**
     * Lista de Entidades -> Lista de DTOs
     */
    public static List<PessoaDto> toDtoList(List<Pessoa> pessoas) {
        if (pessoas == null) return List.of();
        return pessoas.stream()
                .map(PessoaMapper::toDto)
                .collect(Collectors.toList());
    }
}